src/utils/Database/LSMTree/Generators.hs

```diff
@@ -619,9 +631,14 @@ newtype KeyForIndexCompact =
   deriving (Eq, Ord, Show)
 
 instance Arbitrary KeyForIndexCompact where
-  arbitrary =
-      fmap KeyForIndexCompact . genRawBytesN
-          =<< QC.sized (\s -> QC.chooseInt (6, s + 6))
+  -- we try to make collisions and close keys more likely (very crudely)
+  arbitrary = KeyForIndexCompact <$> frequency
+      [ (5, genRawBytesN =<< QC.sized (\s -> QC.chooseInt (6, s + 6)))
+      , (1, do
+          lastByte <- QC.chooseBoundedIntegral (0, 3)
+          return (RB.pack ([1,3,3,7,0] <> [lastByte]))
+        )
```

Comment:
This generates only 3 unique keys, which seems to be a little on the small side. How about something like this:

```suggestion
      , (1, do
          first4Bytes <- serialiseKey . getSmall <$> QC.arbitrary @Word64 
          then2Bytes <- QC.arbitrary @Word16
          otherBytes <- QC.arbitrary
          pure $ first4Bytes <> then2Bytes <> otherBytes
        )
```

Reply:
It's close to what I used before in the `CompactIndex` tests

Reply:
I get zero collisions with that. In the compact index tests, the size parameter is increased a lot, which makes collisions more likely, but ideally we already see some of them at small sizes. We also want multiple collisions, e.g. to test resolving >2 entries in run merging or lookups.

Having a single fixed prefix under which collisions appear is a bit limited, I know, but maybe we can at least have it such that potentially more than 4 values are possible, but some of them are more likely than others (non-uniform distribution).

I just tried this:

```haskell
  arbitrary = KeyForIndexCompact <$> frequency
      [ (6, genRawBytesN =<< QC.sized (\s -> QC.chooseInt (6, s + 6)))
      , (1, do
          lastByte <- QC.sized $ skewedWithMax . fromIntegral
          return (RB.pack ([1,3,3,7,0] <> [lastByte]))
        )
      ]
      where
        -- generates a value in range from 0 to ub, but skewed
        skewedWithMax ub0 = do
          ub1 <- QC.chooseBoundedIntegral (0, ub0)
          ub2 <- QC.chooseBoundedIntegral (0, ub1)
          QC.chooseBoundedIntegral (0, ub2)
```

Allows for last bytes up to max size (usually 100 I think), but at small size limits that range and generally heavily skews towards smaller values. With that we get:

```
        10.1% has key with >2 collisions
         3.2% >10% of keys collide
```

Feels a bit hacky, but do you think that's better?

Reply:
Sounds good, we can use this for now

Reply:
Different idea that I thought of just now, which we don't have to do in this PR: if it's interactions between two separately generated keys that we care about, then maybe we should generate a collection of keys for the compact index instead, so we can precisely decide when to generate two colliding keys or not. Something like:

```haskell
newtype KeysForIndexCompact = KeysForIndexCompact { getKeysForCompactIndex :: [SerialisedKey] }
instance Arbitrary KeysForIndexCompact where ...
```

Reply:
Added it.